### PR TITLE
added request validation for urls with and without ports

### PR DESCRIPTION
**Pull request by eshanholtz**
PR for DI-318

### CODE REVIEW

File: src/main/java/com/twilio/security/RequestValidator.java [hunk 1]
```diff
@@ -107,4 +113,23 @@ private boolean secureCompare(String a, String b) {
...
+        } catch (Exception e) {
+            return "";
+        }
+    }
+    
+    private String addPort(String url) {
+        try {
+            URI parsedUrl = new URI(url);
+            String port = Objects.equals(parsedUrl.getScheme(), "https") ? "443" : "80";
+            return (parsedUrl.getPort()) != -1 ? url : url.replace(".com", ".com:" + port);
```
Comment (childish-sambino) [inline src/main/java/com/twilio/security/RequestValidator.java]: Is there a URI builder in Java that can help here? Something like `https://www.complete.com` would not turn out well.

File: src/main/java/com/twilio/security/RequestValidator.java [hunk 2]
```diff
@@ -26,8 +27,13 @@ public RequestValidator(String authToken) {
     }
 
     public boolean validate(String url, Map<String, String> params, String expectedSignature) {
-        String signature = getValidationSignature(url, params);
-        return secureCompare(signature, expectedSignature);
+        // check signature of url with and without port, since sig generation on back
+        // end is inconsistent
+        String signatureWithPort= getValidationSignature(addPort(url), params);
```
Comment (childish-sambino) [inline src/main/java/com/twilio/security/RequestValidator.java]: nit: space before `=`

File: src/main/java/com/twilio/security/RequestValidator.java [hunk 3]
```diff
@@ -107,4 +113,33 @@ private boolean secureCompare(String a, String b) {
         return mismatch == 0;
     }
 
+    private String removePort(String url) {
+        try {
+            URI parsedUrl = new URI(url);
+            return (parsedUrl.getPort()) == -1 ? url : url.replaceFirst(":\\d+", "");
+        } catch (Exception e) {
+            return "";
```
Comment (childish-sambino) [inline src/main/java/com/twilio/security/RequestValidator.java]: Just return the `url`?

File: src/main/java/com/twilio/security/RequestValidator.java (hunk 1 above)
Comment (eshanholtz) [reply src/main/java/com/twilio/security/RequestValidator.java]: fixed. Thanks for pointing that out!

File: src/main/java/com/twilio/security/RequestValidator.java [hunk 4]
```diff
@@ -107,4 +113,33 @@ private boolean secureCompare(String a, String b) {
         return mismatch == 0;
     }
 
+    private String removePort(String url) {
+        try {
+            URI parsedUrl = new URI(url);
+            return (parsedUrl.getPort()) == -1 ? url : url.replaceFirst(":\\d+", "");
```
Comment (childish-sambino) [inline src/main/java/com/twilio/security/RequestValidator.java]: Doesn't work with this: `https://sambino:1password@www.mysite.com:456` (I use very complex passwords)

Maybe add a function like this:
```
private String updatePort(URI url, int newPort) {
    return new URI(
        url.getScheme(),
        url.getUserInfo(),
        url.getHost(), 
        newPort,
        url.getPath(),
        url.getQuery(),
        url.getFragment()).toString();
}
```

Then you can use it for both adding and removing.

**Pull request merged**